event-schema/event_schema.go

```diff
@@ -617,149 +635,184 @@ func getPrivateDataJSON(schemaHash string) []byte {
+
+	schemaVersionsInCache := make([]*SchemaVersionT, 0)
+	for _, sv := range updatedSchemaVersions {
+		schemaVersionsInCache = append(schemaVersionsInCache, sv)
+	}
+
+	if len(updatedEventModels) == 0 && len(schemaVersionsInCache) == 0 {
+		return nil
+	}
+
+	flushDBHandle := createDBConnection()
+	defer flushDBHandle.Close()
+
+	txn, err := flushDBHandle.Begin()
+	if err != nil {
+		return fmt.Errorf("unable to begin txn to flush event schemas, err: %w", err)
+	}
+
+	if len(updatedEventModels) > 0 {
+		if err := flushEventModels(ctx, txn, updatedEventModels); err != nil {
+			txn.Rollback()
+			return fmt.Errorf("unable to flush event models to db: err: %w", err)
 		}
+	}
 
-		if len(updatedEventModels) == 0 && len(schemaVersionsInCache) == 0 {
-			manager.eventModelLock.Unlock()
-			manager.schemaVersionLock.Unlock()
-			continue
+	if len(schemaVersionsInCache) > 0 {
+		if err := flushSchemaVersions(ctx, txn, schemaVersionsInCache); err != nil {
+			txn.Rollback()
+			return fmt.Errorf("unable to flush schema versions to db, err: %w", err)
 		}
+	}
 
-		flushDBHandle = createDBConnection()
+	err = txn.Commit()
+	if err != nil {
+		return fmt.Errorf("unable to commit txn to update event models and schema versions, err: %w", err)
+	}
 
-		txn, err := flushDBHandle.Begin()
-		assertError(err)
+	updatedEventModels = make(map[string]*EventModelT)
+	updatedSchemaVersions = make(map[string]*SchemaVersionT)
+	toDeleteEventModelIDs = []string{}
+	toDeleteSchemaVersionIDs = []string{}
 
-		// Handle Event Models
-		if len(updatedEventModels) > 0 {
-			eventModelIds := make([]string, 0, len(updatedEventModels))
-			for _, em := range updatedEventModels {
-				eventModelIds = append(eventModelIds, em.UUID)
-			}
+	return nil
+}
 
-			deleteOldEventModelsSQL := fmt.Sprintf(`DELETE FROM %s WHERE uuid IN ('%s')`, EVENT_MODELS_TABLE, strings.Join(eventModelIds, "', '"))
-			_, err := txn.Exec(deleteOldEventModelsSQL)
-			assertTxnError(err, txn)
+func flushSchemaVersions(ctx context.Context, txn *sql.Tx, schemaVersionsInCache []*SchemaVersionT) error {
 
-			if len(toDeleteEventModelIDs) > 0 {
-				archiveOldEventModelsSQL := fmt.Sprintf(`UPDATE %s SET archived=%t WHERE uuid IN ('%s')`, EVENT_MODELS_TABLE, true, strings.Join(toDeleteEventModelIDs, "', '"))
-				_, err := txn.Exec(archiveOldEventModelsSQL)
-				assertTxnError(err, txn)
+	versionIDs := make([]string, 0, len(schemaVersionsInCache))
+	for uid := range updatedSchemaVersions {
+		versionIDs = append(versionIDs, uid)
+	}
 
-				archiveVersionsForArchivedModelsSQL := fmt.Sprintf(`UPDATE %s SET archived=%t WHERE event_model_id IN ('%s')`, SCHEMA_VERSIONS_TABLE, true, strings.Join(toDeleteEventModelIDs, "', '"))
-				_, err = txn.Exec(archiveVersionsForArchivedModelsSQL)
-				assertTxnError(err, txn)
-			}
+	deleteOldVersionsSQL := fmt.Sprintf(`DELETE FROM %s WHERE uuid IN ('%s')`, SCHEMA_VERSIONS_TABLE, strings.Join(versionIDs, "', '"))
+	_, err := txn.Exec(deleteOldVersionsSQL)
```

Comment:
Why not use `ExecContext` instead? It looks like `flushSchemaVersions` does have a context but nobody is using it.

Reply:
I should have done that. Thanks for pointing out ! 👍🏼